### PR TITLE
feat: Prevent user from closing the tab/browser window with unsaved query text #226

### PR DESCRIPTION
**Pull request by vitshev**
#226

### CODE REVIEW

File: packages/ui/src/ui/pages/query-tracker/module/query/reducer.ts [hunk 1]
```diff
@@ -89,12 +92,15 @@ export function reducer(state = initState, action: Actions): QueryState {
             };
         }
         case SET_QUERY_PATCH: {
```
Comment (vitshev) [inline packages/ui/src/ui/pages/query-tracker/module/query/reducer.ts:94]: It seems to me that those two actions SET_QUERY_PATCH/SET_QUERY cover all of our cases.

File: packages/ui/src/ui/hooks/use-before-unload.ts [hunk 2]
```diff
@@ -0,0 +1,18 @@
+import {useEffect} from 'react';
+
+const beforeUnloadHandler = (e: BeforeUnloadEvent) => {
+    e.preventDefault();
+
+    // Included for legacy support, e.g. Chrome/Edge < 119
+    e.returnValue = true;
+};
+
+export const useBeforeUnload = ({shouldListen}: {shouldListen: boolean}) => {
```
Comment (ma-efremoff) [inline packages/ui/src/ui/hooks/use-before-unload.ts:10]: Probably it is better to name the hook `usePreventUnload`

Comment (vitshev) [reply packages/ui/src/ui/hooks/use-before-unload.ts]: renamed it

File: packages/ui/src/ui/pages/query-tracker/module/query/reducer.ts (hunk 1 above)
Comment (ma-efremoff) [reply packages/ui/src/ui/pages/query-tracker/module/query/reducer.ts:94]: All changed parts of the state that received from api should prevent unload. I nearly sure that some data for SET_QUERY_PARAMS comes from api. Could you please check what parts of the state comes from api?

Comment (vitshev) [reply packages/ui/src/ui/pages/query-tracker/module/query/reducer.ts:94]: It seems that SET_QUERY_PARAMS only works with the GET parameters of the page.

**Review by ma-efremoff — APPROVED** [review]

**Review by ma-efremoff — APPROVED** [review]

**Pull request merged**